ts/sagas/authentication.ts

```diff
@@ -0,0 +1,170 @@
+/**
+ * A collection of sagas to manage the Authentication.
+ */
+import { NavigationActions, NavigationState } from "react-navigation";
+import { Effect } from "redux-saga";
+import { call, fork, put, select, take, takeLatest } from "redux-saga/effects";
+
+import { PublicSession } from "../../definitions/backend/PublicSession";
+import { BackendClient, BasicResponseTypeWith401 } from "../api/backend";
+import { apiUrlPrefix } from "../config";
+import ROUTES from "../navigation/routes";
+import {
+  authenticationCompleted,
+  IdpSelected,
+  LoginSuccess,
+  logout,
+  sessionLoadFailure,
+  sessionLoadRequest,
+  sessionLoadSuccess,
+  startAuthentication
+} from "../store/actions/authentication";
+import {
+  AUTHENTICATION_COMPLETED,
+  IDP_SELECTED,
+  LOGIN_SUCCESS,
+  SESSION_EXPIRED,
+  SESSION_LOAD_REQUEST,
+  SESSION_LOAD_SUCCESS,
+  START_AUTHENTICATION
+} from "../store/actions/constants";
+import { navigationRestore } from "../store/actions/navigation";
+import { sessionTokenSelector } from "../store/reducers/authentication";
+import { navigationStateSelector } from "../store/reducers/navigation";
+import { SessionToken } from "../types/SessionToken";
+
+/**
+ * Load session info from the Backend
+ */
+export function* loadSession(): IterableIterator<Effect> {
+  // Gte the SessionToken from the store
+  const sessionToken: SessionToken | undefined = yield select(
+    sessionTokenSelector
+  );
+
+  if (sessionToken) {
+    const backendClient = BackendClient(apiUrlPrefix, sessionToken);
+
+    // Call the Backend service
+    const response:
+      | BasicResponseTypeWith401<PublicSession>
+      | undefined = yield call(backendClient.getSession, {});
+
+    if (!response || response.status !== 200) {
+      // We got a error, send a SESSION_LOAD_FAILURE action
+      const error: Error = response ? response.value : Error();
+
+      yield put(sessionLoadFailure(error));
+    } else {
+      // Ok we got a valid response, send a SESSION_LOAD_SUCCESS action
+      yield put(sessionLoadSuccess(response.value));
+    }
+  } else {
+    // No SessionToken we can't send a SESSION_LOAD_FAILURE action
+    yield put(sessionLoadFailure(Error()));
+  }
+}
+
+/**
+ * Listen for SESSION_EXPIRED.
+ * This saga:
+ * 1) Save the navigation state
+ * 2) Allow the user to re-authenticate
+ * 3) Restore the navigation state to bring the user back to the screen he was
+ *
+ * NOTE: This saga manage SESSION_EXPIRED one by one, this means that if more than one SESSION_EXPIRED
+ * get dispatched at the same time only one get served.
```

Comment:
don't the actions get queued in the channel? meaning that `watchSessionExpired` gets executed two times in sequence?

Reply:
We are NOT using `channel` nor `fork`, we are using a simple `take` so this is the behavior:

1) the saga start listening for a SESSION_EXPIRED action
2) the SESSION_EXPIRED arrives and the rest of the saga body start to be excuted, while in this state the saga is not listening for another SESSION_EXPIRED, so if another SESSION_EXPIRED arrives it will not be handled.
3) the user complete the Authentication, we start again the while loop and start to listen again for SESSION_EXPIRED

Reply:
I see, so when action `A` gets dispatched, unless there is a `take` blocked waiting for action `A`, the action gets discarded?

Reply:
Exactly